micrometer-core/src/main/java/io/micrometer/core/aop/MetricTag.java

```diff
@@ -0,0 +1,64 @@
+/**
+ * Copyright 2022 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.micrometer.core.aop;
+
+import io.micrometer.common.annotation.NoOpTagValueResolver;
+import io.micrometer.common.annotation.TagValueResolver;
+
+import java.lang.annotation.*;
+
+/**
+ * There are 3 different ways to add tags to a span. All of them are controlled by the
+ * annotation values. Precedence is:
+ * <p>
+ * try with the {@link TagValueResolver} bean if the value of the bean wasn't set, try to
+ * evaluate a SPEL expression if there’s no SPEL expression just return a
+ * {@code toString()} value of the parameter
```

Comment:
JavaDoc needs to be updated for metrics (it mentions span and SPEL and beans). It would also be good to make a warning that this should not be used with parameters that are going to result in high cardinality values (e.g. user-provided input) and there is no protection provided out-of-the-box for if it is used in such a way.